src/System.Threading.Thread/src/System/Threading/Thread.cs

```diff
@@ -16,11 +16,14 @@ public sealed partial class Thread : CriticalFinalizerObject
         [ThreadStatic]
         private static Thread t_currentThread;
 
+        [ThreadStatic]
+        private static IPrincipal t_principal;
+        private static AsyncLocal<IPrincipal> s_asyncLocalPrincipal;
```

Comment:
@jkotas, I think you'd requested this split?  Was that for performance reasons or for functional behavior?  If the latter, ok.  If the former, I wonder if we'd be better served just having s_asyncLocalPrincipal, which could still be lazily initialized.  It's a little hard to follow the logic about how/when t_principal gets set, and that would be cleaned up if we didn't have it at all.

Reply:
It wasn't requested by @jkotas I saw this on CurrentCulture and I thought to use same "style" https://github.com/dotnet/corefx/blob/5e3fe25d1236cd0192dd966f7286fe008d5cf875/src/Common/src/CoreLib/System/Globalization/CultureInfo.cs#L125

Reply:
The Principal property should not be used on performance critical paths. It should be ok to not cache it in a thread local, and always go through async-local getter.

Reply:
ok I'll update implementation with direct access to `AsyncLocal` value getter, thank's!

Reply:
done